crates/rsonpath-lib/src/query/automaton.rs

```diff
@@ -18,8 +18,89 @@ pub struct Automaton<'q> {
     states: Vec<StateTable<'q>>,
 }
 
+/// Represent the distinct methods of moving on a match between states.
+#[derive(Debug, Copy, PartialEq, Clone, Eq)]
+pub enum TransitionLabel<'q> {
+    /// Wraps a textual field [`Label`] in a JSON object.
+    ObjectMember(&'q Label),
+    /// Wraps an array index [`NonNegativeArrayIndex`] in a JSON object.
+    ArrayIndex(NonNegativeArrayIndex),
+}
+
+impl<'q> TransitionLabel<'q> {
+    /// Return a display object with a UTF8 representation of this label.
+    #[must_use]
+    #[inline(always)]
+    pub fn display(&self) -> impl Display + 'q {
+        match self {
+            // This isn't just calling label.display() because the compile makes opaque types unique at the declaration site.
+            // So, the one for display() that is defined above would not equate to Label's def.
+            TransitionLabel::ObjectMember(label) => format!("{}", label.display()),
+            TransitionLabel::ArrayIndex(index) => format!("{index}"),
+        }
+    }
+
+    ///Return the textual [`Label`] being wrapped if so.  Returns [`None`] otherwise.
+    #[must_use]
+    #[inline(always)]
+    pub fn get_label(&self) -> Option<&Label> {
+        match self {
+            TransitionLabel::ObjectMember(l) => Some(l),
+            TransitionLabel::ArrayIndex(_) => None,
+        }
+    }
+
+    /// Consumes the [`TransitionLabel`] and gives the wrapped [`Label`], if so.  Returns [`None`] otherwise.
+    #[must_use]
+    #[inline(always)]
+    pub fn get_label_owned(self) -> Option<&'q Label> {
+        match self {
+            TransitionLabel::ObjectMember(l) => Some(l),
+            TransitionLabel::ArrayIndex(_) => None,
+        }
+    }
```

Comment:
Maybe I'm missing something, but why would we need both of those impls? We could have a single `get_label(&self) -> Option<&'q Label>`. `'q` already necessarily outlives a `TransitionLabel<'q>`, so we don't need to consume it.

Reply:
Also: seems weird (from a public API perspective) to not also have an analogous `get_array_index` that returns `Option<NonNegativeArrayIndex>`.

Reply:
> Maybe I'm missing something, but why would we need both of those impls? We could have a single `get_label(&self) -> Option<&'q Label>`. `'q` already necessarily outlives a `TransitionLabel<'q>`, so we don't need to consume it.
That's definitely a more idiomatic solution than provided here.

> Also: seems weird (from a public API perspective) to not also have an analogous `get_array_index` that returns `Option<NonNegativeArrayIndex>`.
That's reasonable to add.

Reply:
> Maybe I'm missing something, but why would we need both of those impls? We could have a single `get_label(&self) -> Option<&'q Label>`. `'q` already necessarily outlives a `TransitionLabel<'q>`, so we don't need to consume it.
A bit rusty on my lifetimes, but yes, this works.  Changed per-se.

> Also: seems weird (from a public API perspective) to not also have an analogous `get_array_index` that returns `Option<NonNegativeArrayIndex>`.
I had this question too, but since there's no current use (of which I'm aware) for the method I skipped it.  Added.